cmd/thanos/query-frontend.go

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package main
+
+import (
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/querier/frontend"
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/run"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/route"
+	"gopkg.in/alecthomas/kingpin.v2"
+
+	v1 "github.com/thanos-io/thanos/pkg/api/queryfrontend"
+	"github.com/thanos-io/thanos/pkg/component"
+	"github.com/thanos-io/thanos/pkg/extflag"
+	"github.com/thanos-io/thanos/pkg/extprom"
+	"github.com/thanos-io/thanos/pkg/prober"
+	"github.com/thanos-io/thanos/pkg/queryfrontend"
+	"github.com/thanos-io/thanos/pkg/queryfrontend/cache"
+	httpserver "github.com/thanos-io/thanos/pkg/server/http"
+)
+
+type queryFrontendConfig struct {
+	http             httpConfig
+	queryRangeConfig queryRangeConfig
+
+	downstreamURL        string
+	compressResponses    bool
+	LogQueriesLongerThan time.Duration
+}
+
+type queryRangeConfig struct {
+	respCacheConfig     extflag.PathOrContent
+	cacheMaxFreshness   time.Duration
+	splitInterval       model.Duration
+	maxRetries          int
+	maxQueryParallelism int
+	maxQueryLength      model.Duration
+}
+
+func (c *queryRangeConfig) registerFlag(cmd *kingpin.CmdClause) {
+	cmd.Flag("query-range.split-interval", "Split queries by an interval and execute in parallel, 0 disables it.").
+		Default("24h").SetValue(&c.splitInterval)
+
+	cmd.Flag("query-range.max-retries-per-request", "Maximum number of retries for a single request; beyond this, the downstream error is returned.").
+		Default("5").IntVar(&c.maxRetries)
+
+	cmd.Flag("query-range.max-query-length", "Limit the query time range (end - start time) in the query-frontend, 0 disables it.").
+		Default("0").SetValue(&c.maxQueryLength)
+
+	cmd.Flag("query-range.max-query-parallelism", "Maximum number of queries will be scheduled in parallel by the frontend.").
+		Default("14").IntVar(&c.maxQueryParallelism)
+
+	cmd.Flag("query-range.response-cache-max-freshness", "Most recent allowed cacheable result, to prevent caching very recent results that might still be in flux.").
```

Comment:
We need better help flag description - in separate PR is ok

Reply:
essentially let's describe why this is needed

Reply:
Make sense, we can also help update the flag description in Cortex as well.